### PR TITLE
Scan Canonicalization: Support for Scans in C++ Backend

### PR DESCRIPTION
**Pull request by fthaler**
## Description

Enables support for scans in the C++ backends.

- Introduces the iterator IR extension nodes `FencilWithTemporaries` and `Temporary` (as wrappers of normal iterator IR; backend support is optional).
- Updates the global temporary transformation to generate a `FencilWithTemporaries and work for Cartesian as well as unstructured grids with (naive) extent analysis.
- Introduces `pass_manager.LiftMode`, allowing to choose different evaluation strategies for lifted function calls.
- Introduces a simple heuristic to decide which lifted function calls should be inlined.
- Introduces new GTIR nodes: `ScanPassDefinition`, `Scan`, `ScanExecution`, and `TemporaryAllocation`.
- Provides lowering of `FencilWithTemporaries` to GTIR and respective C++ code generation.

## Requirements

Before submitting this PR, please make sure:

- [x] You have run the code checks, tests and documentation build successfully
- [x] All fixes and all new functionality are tested and documentation is up to date
- [x] You looked at the [review checklist](https://github.com/GridTools/gt4py/blob/master/REVIEW_CHECKLIST.txt)

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

File: src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py [hunk 1]
```diff
@@ -188,46 +249,148 @@ def visit_FunCall(self, node: itir.FunCall, **kwargs: Any) -> Node:
...
+        backend = Backend(domain=self.visit(node.domain, stencil=node.stencil, **kwargs))
+        if self._is_scan(node.stencil):
+            scan_id = UIDs.sequential_id(prefix="_scan")
+            assert isinstance(node.stencil, itir.FunCall)
+            scan_lambda = self.visit(node.stencil.args[0], **kwargs)
+            forward = self._bool_from_literal(node.stencil.args[1])
+            scan_def = ScanPassDefinition(
+                id=scan_id, params=scan_lambda.params, expr=scan_lambda.expr, forward=forward
+            )
+            extracted_functions.append(scan_def)
```
Comment (havogt) [inline src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py:297]: I am wondering if we should run this extraction as a separate pass on itir.

**Review by havogt — CHANGES_REQUESTED** [review]
I only reviewed a few files yet. Will continue on Monday.

File: src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py [hunk 2]
```diff
@@ -30,18 +31,38 @@
...
+            "float32": "float",
+            "float64": "double",
+            "int": "int",
+            "int32": "std::int32_t",
+            "int64": "std::int64_t",
+            "bool": "bool",
+            "axis_literal": None,  # TODO: domain?
+        }[t]
+    except KeyError:
+        raise TypeError(f"Unsupported type '{t}'") from None
```
[the remark this answers is not in the document]
Comment (fthaler) [reply src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py:63]: I wanted to use `functional.fencil_processors.source_modules.cpp_gen.render_python_type`, but then realized that this actually maps from real Python types and not from strings, so it’s not usable here. So this is left for another cleanup I guess.

File: tests/functional_tests/cpp_backend_tests/tridiagonal_solve_driver.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+constexpr inline auto b = [](auto...) { return 3; };
+constexpr inline auto c = [](auto...) { return 1; };
+constexpr inline auto d = [](int ksize) {
+  return [kmax = ksize - 1](auto... indices) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-value"
+    GT_NVCC_DIAG_PUSH_SUPPRESS(174)
+    int k = (..., indices);
+#pragma GCC diagnostic pop
+    GT_NVCC_DIAG_POP_SUPPRESS(174)
```
[the remark this answers is not in the document]
Comment (fthaler) [reply tests/functional_tests/cpp_backend_tests/tridiagonal_solve_driver.cpp:24]: Sure, this was directly copied from the GT repo which uses the same function for Cartesian and unstructured.

File: src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py (hunk 1 above)
Comment (fthaler) [reply src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py:404]: Yes, this file clearly needs a huge cleanup. But not in this PR, because merging with existing code changes would probably be even more painful then.

File: src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py [hunk 4]
```diff
@@ -188,46 +249,148 @@ def visit_FunCall(self, node: itir.FunCall, **kwargs: Any) -> Node:
...
+                output=Literal(value="0", type="int"),
+                inputs=[Literal(value=str(i + 1), type="int") for i, _ in enumerate(node.inputs)],
+                init=self.visit(node.stencil.args[2], **kwargs),
+            )
+            return ScanExecution(
+                backend=backend,
+                scans=[scan],
+                args=[self.visit(node.output, **kwargs)] + self.visit(node.inputs),
+                axis=SymRef(id=kwargs["column_axis"].value),
+            )
```
[the remark this answers is not in the document]
Comment (fthaler) [reply src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py:416]: I would propose to cleanup (everything is quite messy now) the whole file in a separate PR after merging and also separate the function extraction there.

File: src/functional/fencil_processors/codegens/gtfn/gtfn_ir.py [hunk 5]
```diff
@@ -67,24 +67,31 @@ class SymRef(Expr):
...
     expr: Expr
 
 
 class FunCall(Expr):
     fun: Expr  # VType[Callable]
-    args: List[Expr]
+    args: list[Expr]
 
 
 class FunctionDefinition(Node, SymbolTableTrait):
```
Comment (havogt) [inline src/functional/fencil_processors/codegens/gtfn/gtfn_ir.py:79]: Renaming this to `StencilDefinition` probably doesn't make sense because it covers also non-stencil functions, right?

File: src/functional/fencil_processors/runners/roundtrip.py [hunk 6]
```diff
@@ -179,9 +176,9 @@ def register_tmp(tmp, domain):
         if not debug:
             pathlib.Path(source_file_name).unlink(missing_ok=True)
 
-    assert isinstance(ir, FencilDefinition)
-    fencil_name = ir.id
-    fencil = getattr(foo, fencil_name + "_wrapper")
+    assert isinstance(ir, (itir.FencilDefinition, FencilWithTemporaries))
+    fencil_name = ir.fencil.id + "_wrapper" if isinstance(ir, FencilWithTemporaries) else ir.id
+    fencil = getattr(foo, fencil_name)
```
Comment (havogt) [inline src/functional/fencil_processors/runners/roundtrip.py:181]: We should rename foo. I guess I approved it in the previous commit 😳

File: src/functional/iterator/transforms/inline_lifts.py [hunk 7]
```diff
@@ -1,8 +1,18 @@
+from collections.abc import Callable
+from typing import Optional
+
 from eve import NodeTranslator
 from functional.iterator import ir
 
 
 class InlineLifts(NodeTranslator):
+    def __init__(self, predicate: Optional[Callable[[int], bool]] = None) -> None:
```
Comment (havogt) [inline src/functional/iterator/transforms/inline_lifts.py:9]: please document the predicate

Comment (havogt) [reply src/functional/iterator/transforms/inline_lifts.py:9]: The type hint is wrong, we are passing a `ir.Expr` not `int`.

File: src/functional/iterator/transforms/pass_manager.py [hunk 8]
```diff
@@ -1,27 +1,49 @@
...
+        predicate = heuristic(ir)
+        return InlineLifts(predicate).visit(ir)
+    assert lift_mode == LiftMode.FORCE_TEMPORARIES
+    return ir
+
+
 def apply_common_transforms(
     ir: ir.Node,
-    use_tmps=False,
+    lift_mode=None,
```
Comment (havogt) [inline src/functional/iterator/transforms/pass_manager.py:33]:
```suggestion
    **,
    lift_mode=None,
```
We should not pass all these options as positional arguments. This might break something, therefore we should do it in a separate PR.

File: tests/functional_tests/cpp_backend_tests/tridiagonal_solve_driver.cpp [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+using namespace literals;
+
+constexpr inline auto a = [](auto...) { return -1; };
+constexpr inline auto b = [](auto...) { return 3; };
+constexpr inline auto c = [](auto...) { return 1; };
+constexpr inline auto d = [](int ksize) {
+  return [kmax = ksize - 1](auto, auto, auto k) {
+    return k == 0 ? 4 : k == kmax ? 2 : 3;
+  };
+};
```
Comment (havogt) [inline tests/functional_tests/cpp_backend_tests/tridiagonal_solve_driver.cpp:21]: reminder to back-port this change to gridtools c++ example

File: src/functional/iterator/transforms/simple_inline_heuristic.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
+from collections.abc import Callable
+
+from eve import NodeVisitor
+from functional.iterator import ir
+
+
+class AnalyzeLifts(NodeVisitor):
```
Comment (havogt) [inline src/functional/iterator/transforms/simple_inline_heuristic.py:7]: Please document the strategy. This just prevents inlining scans?

Comment (havogt) [reply src/functional/iterator/transforms/simple_inline_heuristic.py:7]: Also document in the right place (in the global_tmps pass) that this is required for scans and why.

File: src/functional/iterator/embedded.py [hunk 11]
```diff
@@ -187,6 +187,11 @@ def or_(a, b):
 
 @builtins.tuple_get.register(EMBEDDED)
 def tuple_get(i, tup):
+    if isinstance(tup, tuple):
+        return tup[i]
+    assert isinstance(tup, (np.ndarray, np.void))
+    if tup.dtype.names:
+        return tup[tup.dtype.names[i]]
```
Comment (havogt) [inline src/functional/iterator/embedded.py:194]: please add a comment that this is to support structured types (if I understand it correctly)

File: src/functional/iterator/embedded.py [hunk 12]
```diff
@@ -187,6 +187,11 @@ def or_(a, b):
...
     return tup[i]
```
Comment (havogt) [inline src/functional/iterator/embedded.py:195]: is this line reachable?

File: src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py [hunk 13]
```diff
@@ -13,10 +13,11 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
 
 
-from typing import Any, Iterable, Type
+from typing import Any, Iterable, Optional, Type, Union
```
Comment (havogt) [inline src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py:16]: Let's put this whole file on the refactoring list to do directly after merge.

File: src/functional/iterator/transforms/global_tmps.py [hunk 14]
```diff
@@ -1,25 +1,159 @@
...
+    hcall = printer._hmerge([node.fencil.id + "("], args, [");"])
+    vcall = printer._vmerge(printer._hmerge([node.fencil.id + "("]), printer._indent(args), [");"])
+    call = printer._optimum(hcall, vcall)
+
+    body = printer._vmerge(*tmps, fencil, call)
+    return printer._vmerge(params, printer._indent(body), ["}"])
+
+
+PrettyPrinter.visit_Temporary = pformat_Temporary  # type: ignore
+PrettyPrinter.visit_FencilWithTemporaries = pformat_FencilWithTemporaries  # type: ignore
```
Comment (havogt) [inline src/functional/iterator/transforms/global_tmps.py:74]: We need to find a better place for these. Let's put a TODO for now and document somewhere (maybe ADR) why we do this IR extension.

**Review by havogt — CHANGES_REQUESTED** [review]
I didn't review the global_tmps pass yet. I let you decide to implement the requested changes now or add a follow up PR for cleanup.

**Comment by havogt** [comment]
A reminder for me to check the result of using this branch in ICON: https://github.com/C2SM/icon-exclaim/pull/35

File: src/functional/fencil_processors/codegens/gtfn/gtfn_ir.py (hunk 5 above)
Comment (fthaler) [reply src/functional/fencil_processors/codegens/gtfn/gtfn_ir.py:79]: Correct, can be anything.

File: src/functional/fencil_processors/runners/roundtrip.py (hunk 6 above)
Comment (fthaler) [reply src/functional/fencil_processors/runners/roundtrip.py]: Fixed you terrible review failure 😂

File: src/functional/iterator/embedded.py (hunk 12 above)
Comment (fthaler) [reply src/functional/iterator/embedded.py:195]: Yes, it’s for fields with tuple dimensions. Clarified in the comments.

File: src/functional/iterator/transforms/simple_inline_heuristic.py (hunk 10 above)
Comment (fthaler) [reply src/functional/iterator/transforms/simple_inline_heuristic.py:7]: It prevents inlining scans and inling of the first lifted functions within a scan. Otherwise we would get scan-only computations. It’s only strictly needed by gtfn, I added comments (to the `heuristic` function).

File: tests/functional_tests/cpp_backend_tests/tridiagonal_solve_driver.cpp (hunk 9 above)
Comment (fthaler) [reply tests/functional_tests/cpp_backend_tests/tridiagonal_solve_driver.cpp:21]: It can’t be backported to GT C++, this simplification was only possible because we only test the Cartesian case here.

File: src/functional/iterator/transforms/simple_inline_heuristic.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def heuristic(node: ir.Node) -> Callable[[ir.Node], bool]:
+    """Create a predicate for the InlineLifts transformation.
+
+    Follows the simple rules:
+    - Don’t inline scans (as there is no efficient way to inline them, also required by some
+      backends, e.g. gtfn)
+    - Don’t inline the first lifted function call within a scan (otherwise, all stencils would get
+      inlined into the scans, leading to reduced parallelism/scan-only computatio)
```
Comment (havogt) [inline src/functional/iterator/transforms/simple_inline_heuristic.py:36]:
```suggestion
      inlined into the scans, leading to reduced parallelism/scan-only computation)
```

File: src/functional/iterator/transforms/pass_manager.py [hunk 16]
```diff
@@ -40,15 +63,14 @@ def apply_common_transforms(
...
+    if lift_mode != LiftMode.FORCE_INLINE:
         assert offset_provider is not None
-        ir = CreateGlobalTmps().visit(
-            ir, offset_provider=offset_provider, register_tmp=register_tmp
-        )
+        ir = CreateGlobalTmps().visit(ir, offset_provider=offset_provider)
+        ir = InlineLifts().visit(ir)
 
     if common_subexpression_elimination:
         ir = CommonSubexpressionElimination().visit(ir)
```
Comment (havogt) [inline src/functional/iterator/transforms/pass_manager.py:76]:
```suggestion
        ir = EtaReduction().visit(ir)
        ir = CommonSubexpressionElimination().visit(ir)
```
fixes the ICON stencil

**Review by havogt — CHANGES_REQUESTED** [review]
still missing the review of global_tmps pass

File: src/functional/iterator/transforms/global_tmps.py [hunk 17]
```diff
@@ -1,25 +1,159 @@
...
+
+
+def split_closures(node: ir.FencilDefinition) -> FencilWithTemporaries:
```
Comment (havogt) [inline src/functional/iterator/transforms/global_tmps.py:77]: Please add a docstring

File: src/functional/iterator/transforms/global_tmps.py [hunk 18]
```diff
@@ -1,25 +1,159 @@
...
+            closures.append(closure)
+            domain = AUTO_DOMAIN
+
+    return FencilWithTemporaries(
+        fencil=ir.FencilDefinition(
+            id=node.id,
+            function_definitions=node.function_definitions,
+            params=node.params
+            + [ir.Sym(id=tmp.id) for tmp in tmps]
+            + [ir.Sym(id="_gtmp_auto_domain")],
```
Comment (havogt) [inline src/functional/iterator/transforms/global_tmps.py:121]: put `_gtmp_auto_domain` in a variable to make refactoring easier

File: src/functional/iterator/transforms/global_tmps.py [hunk 19]
```diff
@@ -1,25 +1,159 @@
...
+            isinstance(arg, ir.FunCall)
+            and isinstance(arg.fun, ir.FunCall)
+            and isinstance(arg.fun.fun, ir.SymRef)
+            and arg.fun.fun.id == "lift"
+        ):
+            assert len(arg.fun.args) == 1
+            ref = ir.SymRef(id=f"_gtmp_{len(tmps)}")
+            tmps.append(ir.Sym(id=ref.id))
+            unlifted = ir.FunCall(fun=arg.fun.args[0], args=arg.args)
+            todos.append((ref, unlifted))
```
Comment (havogt) [inline src/functional/iterator/transforms/global_tmps.py:93]: I hate late binding, is it acceptable to use it in that way because you are supposed to understand it?

File: src/functional/iterator/transforms/global_tmps.py [hunk 20]
```diff
@@ -67,78 +195,152 @@ def _extend_domain(domain: ir.FunCall, offset_provider, shifts):
...
+            params=node.fencil.params[:-1],
             closures=list(reversed(closures)),
-        )
+        ),
+        params=node.params,
+        tmps=node.tmps,
+    )
+
+
+def update_unstructured_domains(node: FencilWithTemporaries, offset_provider):
```
Comment (havogt) [inline src/functional/iterator/transforms/global_tmps.py:234]: there is no test for this, should we add one or does this need to change anyway in the near future?

File: src/functional/iterator/transforms/global_tmps.py [hunk 21]
```diff
@@ -1,25 +1,159 @@
...
-from functional.iterator import ir
+from eve import Coerced, NodeTranslator
+from eve.traits import SymbolTableTrait
+from functional.iterator import ir, type_inference
+from functional.iterator.pretty_printer import PrettyPrinter
 from functional.iterator.runtime import CartesianAxis
 from functional.iterator.transforms.collect_shifts import CollectShifts
+from functional.iterator.transforms.eta_reduction import EtaReduction
 from functional.iterator.transforms.popup_tmps import PopupTmps
+from functional.iterator.transforms.prune_closure_inputs import PruneClosureInputs
```
Comment (havogt) [inline src/functional/iterator/transforms/global_tmps.py:13]: Let's describe the strategy with domain update here:
- introduce magic value `_gtmp_auto_domain` in extraction pass
- then fix the domain afterwards

**Review by havogt — CHANGES_REQUESTED** [review]
only a few minor things left

File: src/functional/iterator/transforms/pass_manager.py (hunk 8 above)
Comment (fthaler) [reply src/functional/iterator/transforms/pass_manager.py:35]: Didn’t break anything, so I forced keyword-only args.

File: src/functional/iterator/transforms/pass_manager.py (hunk 16 above)
Comment (fthaler) [reply src/functional/iterator/transforms/pass_manager.py:79]: I put the EtaReduction outside of the if-condition as it is required by the gtfn backend. Could you double-check that all ICON stuff still works?

File: src/functional/iterator/tracing.py [hunk 22]
```diff
@@ -148,7 +148,7 @@ def make_node(o):
         return list(make_node(arg) for arg in o)
     if o is None:
         return NoneLiteral()
-    if isinstance(o, iterator.runtime.FundefDispatcher):
+    if hasattr(o, "fun"):
```
Comment (fthaler) [inline src/functional/iterator/tracing.py:151]: Change required to support unapplied builtin functions (can be introduced by the new EtaReduction pass).

File: src/functional/iterator/transforms/pass_manager.py (hunk 16 above)
Comment (havogt) [reply src/functional/iterator/transforms/pass_manager.py:79]: Currently it compiled, but failed verification. I compared the generated code for one stencil and there were only name changes. Rerunning ICON test now.

File: src/functional/iterator/transforms/global_tmps.py (hunk 19 above)
Comment (fthaler) [reply src/functional/iterator/transforms/global_tmps.py]: I don’t know, it of course requires people to understand Python’s scoping. When explicitly passing the lists, it requires people to understand Python’s pass-by-object-reference. So I chose the former as the latter further requires passing unnecessary arguments.

**Comment by fthaler** [comment]
@havogt I implemented a much more robust extent analysis in https://github.com/fthaler/gt4py/pull/2
The changes are a bit more heavy than probably expected as it required a new shift/offset analysis. We can either first merge this one here and I will open a separate PR, or you can review the other one and merge into this one. I just kept it separate for now to simplify the review process.

**Review by havogt — APPROVED** [review]
Can you summarize the major things of this PR and its limitations in the commit message
